reposerver/repository/repository.go

```diff
@@ -754,8 +755,103 @@ func (s *Service) runManifestGenAsync(ctx context.Context, repoRoot, commitSHA,
 						}
 					}
 				}
-			}
-		}
+			} // if q.ApplicationSource.Helm != nil {
+
+			// Before generating the manifests, copy any structure from other
+			// This is done for all types of sources, not just helm.
+
+			// If the applicationsource has declared any subtrees to be copied
+			for _, copySpec := range q.ApplicationSource.From {
+				// First resolve the referenced source repo
+				// ref is of type RefTarget
+
+				refVar := copySpec.Ref
+				refSourceMapping, ok := q.RefSources[copySpec.Ref]
+				if !ok {
+					if len(q.RefSources) == 0 {
+						ch.errCh <- fmt.Errorf("source in from referenced %q, but no source has a 'ref' field defined", refVar)
+					}
+					refKeys := make([]string, 0)
+					for refKey := range q.RefSources {
+						refKeys = append(refKeys, refKey)
+					}
+					ch.errCh <- fmt.Errorf("source in from referenced %q, which is not one of the available sources (%s)", refVar, strings.Join(refKeys, ", "))
```

Comment:
Could we add `q.RefSources` directly into logs instead of interating over it and creating a new list `refKeys` for logging?

Reply:
@ishitasequeira the same code is at L689 (I've copied it from there). I think removing both iterations, and logging the complete list of sources at the begining at INFO would be preferred. Or at DEBUG, depending on how much default visibility do we want to have in the logs.

Reply:
@ishitasequeira  And I think the goal of the iteration is string conversion. When directly feeding the `q.RefSources` into `strings.Join()` the implicit type conversion operator is not kicking int:
```
cannot use q.RefSources (variable of type map[string]*"github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1".RefTarget) as []string value in argument to strings.Join
```

Reply:
`20f3bf4c` should be fixing this.